### DISCUSSION TITLE
Build failed: "The type or namespace name 'Epic' could not be found"

### DISCUSSION THREAD

**Discussion by Jade-Harleyy** [FakeFishGames/Barotrauma#13726] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

When trying to build Barotrauma in Visual Studio, I recieve the following error:
**Error** **CS0246** - The type or namespace name 'Epic' could not be found (are you missing a using directive or an assembly reference?)

The error appears for the following files:
- Libraries\BarotraumaLibs\EosInterfacePrivate\InterfaceImpl\
  - Core\
    - CorePrivate.cs
  - Friends\
    - FriendsPrivate.cs
    - PresencePrivate.cs
  - IdAndAuth\
    - EgsIdTokenPrivate.cs
    - IdQueriesPrivate.cs
    - LoginPrivate.cs
  - P2P\
    - P2PSocketPrivate.cs
  - Sessions\
    - OwnedSessionsPrivate.cs
    - RemoteSessionsPrivate.cs
  - Util\
    - ResultExtension.cs

### Reproduction steps

1. Clone the Barotrauma repository's master branch.
2. Open WindowsSolution.sln in Visual Studio 2022.
3. Set any build configuration.
4. Build solution.

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Other

### -

Building Solution

### Version

v1.3.0.2

### Which operating system did you encounter this bug on?

Windows

**Comment by Jade-Harleyy** [comment]
While this doesn't prevent me from making pull requests, it *does* prevent me from actually testing any code I write, which I would like to be fixed.
Please notify me if any workarounds are found! Thank you.

**Reply by Jade-Harleyy** [reply]
Found a workaround:
Unload Libraries\BarotraumaLibs\EosInterfacePrivate\EosInterface.Implementation.Win64.csproj in the solution explorer. :3

**Comment by dossy** [comment]
I just ran into this issue, and this is due to the dependency on the Epic Online Services SDK (EOS-SDK).

You will need to sign in to the [Epic Games Developer Portal](https://dev.epicgames.com/portal), create your Organization and Product if you haven't already, and then use the "Downloads & Release notes" button to download the **C# SDK** version 1.16.1-CL30020700, 

The reason to use this specific older version is because [Epic broke backwards compatibility when releasing 1.16.2-CL32273396](https://dev.epicgames.com/docs/epic-online-services/release-notes#1162-cl32273396---2024-mar-19):

> **New (breaking change):** In the ReceivePacket method, two out parameters have changed to ref parameters. This allows the arguments to be buffers and so avoids constantly creating new objects.

Once you've downloaded the correct version of the EOS C# SDK, unzip only the SDK files from the .zip into `Libraries/BarotraumaLibs/EosInterfacePrivate` like this:

```shell
cd Libraries/BarotraumaLibs/EosInterfacePrivate
unzip /path/to/your/EOS-SDK-CSharp-41373641-v1.17.0.zip SDK/*
mv SDK EOS-SDK
ln -sf Bin EOS-SDK/Redist
```

I originally made the mistake of leaving the `Samples` folder from the SDK in, and the .NET build tried to build them all and threw errors.  You only need the `SDK` folder from the SDK, and it needs to be named `EOS-SDK` in the `Libraries/BarotraumaLibs/EosInterfacePrivate` folder.  I also needed to create the symbolic link between `EOS-SDK/Bin` and `EOS-SDK/Redist` so that the build could find `EOS-SDK/Redist/libEOSSDK-Mac-Shipping.dylib`.

Against the code in the current `master` branch as of this writing ([6c1b18a](https://github.com/FakeFishGames/Barotrauma/tree/6c1b18a56d17ec55c84840d8ebffd24f983ada62)), using this version of the SDK will get you to the point where the build will still fail with these 4 errors:

```
.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/InterfaceImpl/Core/CorePrivate.cs(79,36): error CS0103: The name 'PlatformInterfaceOptionsPrivate' does not exist in the current context [.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.MacOS.csproj]
.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/InterfaceImpl/IdAndAuth/OwnershipPrivate.cs(29,17): error CS0103: The name 'AudienceItemId' does not exist in the current context [.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.MacOS.csproj]
.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/InterfaceImpl/Friends/FriendsPrivate.cs(111,33): error CS0103: The name 'PlatformInterfaceOptionsPrivate' does not exist in the current context [.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.MacOS.csproj]
.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/InterfaceImpl/IdAndAuth/OwnershipPrivate.cs(93,30): error CS0103: The name 'AudienceItemId' does not exist in the current context [.../Barotrauma/Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.MacOS.csproj]
```

I suspect these need to be defined in a file that hasn't been checked into the git repo because they contain sensitive/private information.  It would have been nice if a `.dist` or `.sample` file was checked into git to make it easy for us to fill in our details or have placeholders so that we can at least get a build successfully built, but that's for another time.

**Comment by leca457** [comment]
As @Jade-Harleyy stated, it is really sufficient to remove aforementioned dependency from the solution file.

Here is a way to do it without using GUI/VS Code:
Linux:
```
dotnet sln LinuxSolution.sln remove Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.Linux.csproj
```
MacOS
```
dotnet sln MacSolution.sln remove Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.MacOS.csproj
```
Windows
```
dotnet sln WindowsSolution.sln remove Libraries/BarotraumaLibs/EosInterfacePrivate/EosInterface.Implementation.Win64.csproj
```
